builtin/logical/database/backend.go

```diff
@@ -228,6 +272,156 @@ func (b *databaseBackend) GetConnection(ctx context.Context, s logical.Storage,
+		}
+
+		// load roles and populate queue with static accounts
+		b.populateQueue(ctx, conf.StorageView)
+
+		// launch ticker
+		go b.runTicker(ctx, conf.StorageView)
+	}
+}
+
+// loadStaticWALs reads WAL entries at backend initialization. WAL entries are
+// created during the normal rotation process to ensure data integrity, in the
+// event that Vault fails to store the results (due to power failure, disk
+// failure, et. al.) If WAL entries for static account rotation exist,
+// attempt to re-set the password for the role given the NewPassword stored in
+// the WAL. If the matching Role does not exist, the Role's LastVaultRotation is
+// newer than the WAL, or the Role does not have a static account, delete the
+// WAL.
+func (b *databaseBackend) loadStaticWALs(ctx context.Context, conf *logical.BackendConfig) error {
+	keys, err := framework.ListWAL(ctx, conf.StorageView)
+	if err != nil {
+		return err
+	}
+	if len(keys) == 0 {
+		b.Logger().Debug("no WAL entries found")
+		return nil
+	}
+
+	// loop through WAL keys and process any rotation ones
+	var merr error
+	for _, walID := range keys {
+		// allow cancellation from context
+		select {
+		case <-ctx.Done():
+			b.Logger().Info("loading WAL entries cancelled")
+			return merr
+		default:
+		}
+
+		walEntry := b.findStaticWAL(ctx, conf.StorageView, walID)
+		if walEntry == nil {
+			continue
+		}
+
+		// load matching role and verify
+		role, err := b.StaticRole(ctx, conf.StorageView, walEntry.RoleName)
+		if err != nil {
+			b.Logger().Warn("error loading role", err)
+			merr = multierror.Append(merr, err)
+			continue
+		}
+
+		if role == nil || role.StaticAccount == nil {
+			b.Logger().Warn("role or static account not found")
+			if err = framework.DeleteWAL(ctx, conf.StorageView, walID); err != nil {
+				b.Logger().Warn("error deleting WAL for role with no static account", err.Error())
+				merr = multierror.Append(merr, err)
+			}
+			continue
+		}
+
+		if role.StaticAccount.LastVaultRotation.After(walEntry.LastVaultRotation) {
+			// role password has been rotated since the WAL was created, so let's
+			// delete this
+			if err = framework.DeleteWAL(ctx, conf.StorageView, walID); err != nil {
+				b.Logger().Warn("error deleting WAL for role with newer rotation date", err.Error())
+				merr = multierror.Append(merr, err)
+			}
+			continue
+		}
+
+		// setStaticAccount which will attempt to set the password and
+		// delete the WAL if successful
+		resp, err := b.setStaticAccount(ctx, conf.StorageView, &setStaticAccountInput{
+			RoleName: walEntry.RoleName,
+			Role:     role,
+			WALID:    walID,
+			Password: walEntry.NewPassword,
+		})
+		if err != nil {
+			// if response contains a WALID, create an item to push to the queue with
+			// a backoff time and include the WAL ID
+			merr = multierror.Append(merr, err)
+			if resp.WALID != "" {
+				// Add their rotation to the queue
+				if err := b.pushItem(&queue.Item{
```

Comment:
Is this necessary if the caller of this function goes on to call `populateQueue`?

Reply:
Maybe so, since this is only on error?

Reply:
Here we add in 60s backoff time due to the error returned. If the same item is added later in a follow-up call in `populateQueue`, that attempt will end in an error because the queue currently rejects items with a duplicate name.